rosbag2_transport/test/rosbag2_transport/record_integration_fixture.hpp

```diff
@@ -50,21 +47,33 @@ class RecordIntegrationTestFixture : public Rosbag2TransportTestFixture
 
   void TearDown() override
   {
+    stop_spinning();
     rclcpp::shutdown();
   }
 
   template<class T>
   void start_async_spin(T node)
   {
-    future_ = std::async(
-      std::launch::async, [node]() -> void {rclcpp::spin(node);});
+    if (exec_ == nullptr) {
+      exec_ = std::make_unique<rclcpp::executors::SingleThreadedExecutor>();
+      exec_->add_node(node);
+      spin_thread_ = std::thread(
+        [this]() {
+          exec_->spin();
+        });
+    } else {
+      throw std::runtime_error("Already spinning a node, can't start a new node spin");
+    }
```

Comment:
```suggestion
    if (exec_ == nullptr) {
      exec_ = std::make_unique<rclcpp::executors::SingleThreadedExecutor>();
      exec_->add_node(node);
      spin_thread_ = std::thread(
        [this]() {
          exec_->spin();
        });
      // Wait for the executor to start spinning in the newly spawned thread to avoid race condition
      // with exec_->cancel()
      using clock = std::chrono::steady_clock;
      auto start = clock::now();
      while (!exec_->is_spinning() && (clock::now() - start) < std::chrono::seconds(5)) {
        std::this_thread::sleep_for(std::chrono::milliseconds(20));
      }
      if (!exec_->is_spinning()) {
        throw std::runtime_error("Failed to start spinning node");
      }
    } else {
      throw std::runtime_error("Already spinning a node, can't start a new node spin");
    }
```

Reply:
Oh!  Fantastic.  Thank you so much for looking into it.  I'm going to test this out.